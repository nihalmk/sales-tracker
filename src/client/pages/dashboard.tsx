import React, { useState } from 'react';
import { useContext } from 'react';
import UserContext from '../components/UserWrapper/UserContext';
import { NextPage } from 'next';
import { Layout } from '../components/Layout/Layout';
import ErrorMessage from '../components/Errors/ErrorMessage';
import Navigation from '../components/Navigation/Navigation';

interface Props {}

const Home: NextPage<Props> = () => {
  const { user } = useContext(UserContext);

  const [selectedMenu, setSelectedMenu] = useState<string>();

  return (
    <Layout hideHeader={false}>
      <div className="container">
        {user?.shop ? (
          <React.Fragment>
            <Navigation selected={setSelectedMenu} />
            {selectedMenu}
          </React.Fragment>
        ) : (
          <ErrorMessage
            error={`You don't have any shop assigned to you. Please contact your admin and get assigned to a Shop`}
          />
        )}
        <style jsx global>{``}</style>
      </div>
    </Layout>
  );
};

export default Home;
