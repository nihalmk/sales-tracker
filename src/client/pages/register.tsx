import React from 'react';
import { NextPage } from 'next';
import { Layout } from '../components/Layout/Layout';
import Register from '../components/User/Register';

interface Props {}

const RegisterPage: NextPage<Props> = () => {

  return (
    <React.Fragment>
      <Layout hideHeader={false}><Register /></Layout>
    </React.Fragment>
  );
};

export default RegisterPage;
