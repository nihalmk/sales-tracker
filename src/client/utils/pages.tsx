export const Pages = {
  INDEX: '/',
  DASHBOARD: '/dashboard',
  LOGIN: '/login',
};

export const AuthenticatedPages: { [key: string]: string } = {
  DASHBOARD: '/dashboard',
  INDEX: '/',
};

export const Titles: { [key: string]: string } = {
  HOME: 'Sales Tracker',
  '/': 'Order Dashboard',
  '/dashboard': 'Order Dashboard',
  '/login': 'Login',
};
