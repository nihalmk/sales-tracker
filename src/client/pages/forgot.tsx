import React, { useState, useEffect } from 'react';
import { NextPage } from 'next';
import { Logo } from '../components/Header/Logo';
import { Alert } from 'tabler-react';
import { useMutation } from '@apollo/react-hooks';
import { RESET_PASSWORD } from '../graphql/mutation/user';
import { useRouter } from 'next/router';
import { clientLogger as logger } from '../utils/logger';

interface Props {}

interface Notification {
  message: string;
  type: string;
}

export const Forgot: NextPage<Props> = () => {
  const [email, setEmail] = useState('');
  const [isLoading, setLoading] = useState(false);
  const [notification, setNotification] = useState(undefined);
  const [sendResetMail, {}] = useMutation(RESET_PASSWORD);
  const router = useRouter();
  const onEmailChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setEmail(event.target.value);
  };

  useEffect(() => {
    if (router.query.from === 'reset') {
      setNotification({
        type: 'danger',
        message:
          'Your password reset link is either invalid or expired. Create a new link below.',
      });
    }
  }, []);

  const sendMail = async () => {
    //validating
    if (!/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(email)) {
      setNotification({
        type: 'danger',
        message: 'Invalid email address',
      });
      return;
    }
    setLoading(true);
    try {
      const res = await sendResetMail({
        variables: {
          email,
        },
      });
      let result = res.data.sendResetPasswordEmail;
      if (result) {
        setNotification({
          type: 'success',
          message:
            'An email with instructions to reset password has been sent.',
        });
      } else {
        setNotification({
          type: 'danger',
          message: 'User not found.',
        });
      }

      setLoading(false);
    } catch (ex) {
      logger.error(ex.Message);
      setLoading(false);
    }
  };

  const getNotification = (notificationObj: Notification): JSX.Element => {
    return (
      <Alert
        id="notification"
        className="alert-align"
        type={notificationObj.type}
      >
        {notificationObj.message}
      </Alert>
    );
  };

  return (
    <div className="col col-login mx-auto">
      <div className="text-center mb-6">
        <Logo setColor />
      </div>
      <div className="card">
        <div className="card-body">
          <div className="card-title text-center">Forgot Password</div>
        </div>
        <div className="card-body p-6">
          {notification && getNotification(notification)}
          <p>
            Enter your email address to receive an email with password reset
            instructions.
          </p>

          <div className="form-group">
            <label className="form-label">Email</label>
            <input
              name="email"
              type="email"
              className="form-control"
              id="email"
              placeholder="Enter email"
              onChange={onEmailChange}
            />
          </div>
          <div className="form-footer">
            <button
              type="button"
              onClick={sendMail}
              className={
                'btn btn-primary btn-block ' + (isLoading && 'btn-loading')
              }
            >
              Send reset email
            </button>
          </div>
        </div>
      </div>
    </div>
  );
};

export default Forgot;
