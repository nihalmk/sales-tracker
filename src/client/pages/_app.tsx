import React from 'react';
import Head from 'next/head';
import App from 'next/app';
import { withApollo } from '../apollo/client';
import 'tabler-react/dist/Tabler.css';
import "react-datepicker/dist/react-datepicker.css";
import UserWrapper from '../components/UserWrapper/UserWrapper';
// import { config } from '@fortawesome/fontawesome-svg-core';
// import '@fortawesome/fontawesome-svg-core/styles.css';
import Router from 'next/router';
import { AuthenticatedPages } from '../utils/pages';

// config.autoAddCss = false;

class MyApp extends App {
  state = {
    loaded: false,
  };

  componentDidMount() {
    this.setState({ loaded: true });
  }

  render(): React.ReactElement {
    const { Component, pageProps } = this.props;

    let path = Router && Router.router && Router.router.route;

    const isAuthenticationNeeded = () => {
      return Object.values(AuthenticatedPages).includes(path);
    };

    return (
      <React.Fragment>
        <Head>
          <title>Shop Tracker</title>
          <link rel="icon" href="/static/favicon.ico" />
        </Head>
        {this.state.loaded &&
          (isAuthenticationNeeded() ? (
            <UserWrapper
              Component={Component}
              pageProps={pageProps}
            ></UserWrapper>
          ) : (
            <Component {...pageProps} />
          ))}
      </React.Fragment>
    );
  }
}

export default withApollo(MyApp);
