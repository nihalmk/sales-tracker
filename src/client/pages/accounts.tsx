import React from 'react';
import { NextPage } from 'next';
import { Layout } from '../components/Layout/Layout';
import Link from 'next/link';
import { Pages } from '../utils/pages';

interface Props {}

const Accounts: NextPage<Props> = () => {
  return (
    <React.Fragment>
      <Layout hideHeader={false}>
        <div className="alert alert-primary" role="alert">
          Please send a mail to <strong>shoptrackerforsales@gmail.com</strong>{' '}
          to receive details regarding purchasing a paid account!
        </div>
        <Link href={Pages.DASHBOARD}>
          <button className="btn btn-primary mt-3">Go to home page</button>
        </Link>
      </Layout>
    </React.Fragment>
  );
};

export default Accounts;
