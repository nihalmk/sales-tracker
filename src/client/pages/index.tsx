import React, { useContext } from 'react';
import { useEffect } from 'react';
import { NextPage } from 'next';
import { clientLogger } from '../utils/logger';
import { useRouter } from 'next/router';
import { Pages } from '../utils/pages';
import Loader from '../components/Loaders/Loader';
import UserContext from '../components/UserWrapper/UserContext';
import AddShop from '../components/Shop/AddShop';
import { Layout } from '../components/Layout/Layout';

interface Props {}

const Home: NextPage<Props> = () => {
  const { user } = useContext(UserContext);

  const shop = user?.shop;

  const router = useRouter();
  useEffect(() => {
    if (shop) {
      clientLogger.log('redirecting to dashboard..');
      router.push(Pages.DASHBOARD);
    }
  }, [shop]);

  return (
    <React.Fragment>
      <Layout hideHeader={false}>{shop ? <Loader /> : <AddShop />}</Layout>
    </React.Fragment>
  );
};

export default Home;
