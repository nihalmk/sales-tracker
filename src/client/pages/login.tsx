import Head from 'next/head';
import { useState, useEffect } from 'react';
import { login } from '../accounts/login';
import { useRouter } from 'next/router';
import { Pages } from '../utils/pages';
import { NextPage } from 'next';
import Loader from '../components/Loaders/Loader';
import ErrorMessage from '../components/Errors/ErrorMessage';
import { Logo } from '../components/Header/Logo';
import { GET_USER } from '../graphql/query/user';
import { useQuery } from '@apollo/react-hooks';
import Link from 'next/link';

interface Props {}

export const Login: NextPage<Props> = function () {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [isLoading, setLoading] = useState(false);
  const [error, setError] = useState('');
  const router = useRouter();

  const currentUser = useQuery(GET_USER, {
    fetchPolicy: 'no-cache',
  });

  const onEmailChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setEmail(event.target.value);
  };

  const onPasswordChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setPassword(event.target.value);
  };

  const onSubmit = async (e: React.SyntheticEvent) => {
    e && e.preventDefault();
    setLoading(true);
    try {
      if (!email || !password) {
        setError('Email/Password required');
        setLoading(false);
        return;
      }
      await login(email, password);
      router.push(Pages.INDEX);
    } catch (e) {
      setError(e.message);
      setLoading(false);
    }
  };

  useEffect(() => {
    if (currentUser?.data?.me) {
      router.push(Pages.DASHBOARD);
    }
  }, [currentUser]);

  if (currentUser.loading || currentUser?.data?.me) {
    return <Loader />;
  }
  return (
    <div className="container">
      <Head>
        <title>Login</title>
        <link rel="icon" href="/favicon.ico" />
      </Head>
      <div className="row">
        <div className="col col-login mx-auto">
          <div className="text-center mb-6">
            <Logo setColor />
          </div>
          <form className="card" action="" method="post" onSubmit={onSubmit}>
            <div className="card-body p-6">
              <div className="card-title text-center">Sign in to your Shop</div>
            </div>
            <div className="card-body p-6">
              <ErrorMessage error={error}></ErrorMessage>
              <div className="form-group">
                <label className="form-label">Email address</label>
                <input
                  name="email"
                  type="email"
                  className="form-control"
                  id="email"
                  placeholder="Enter email"
                  onChange={onEmailChange}
                />
              </div>
              <div className="form-group">
                <label className="form-label">
                  Password
                  {/* <a className="float-right small" href="/forgot">
                    I forgot my password
                  </a> */}
                </label>
                <input
                  name="password"
                  type="password"
                  className="form-control"
                  id="userPassword"
                  placeholder="Password"
                  onChange={onPasswordChange}
                />
              </div>
              <Link href="/forgot">Forgot Password?</Link>
              <div className="form-footer">
                <button
                  type="button"
                  onClick={onSubmit}
                  className={
                    'btn btn-primary btn-block ' + (isLoading && 'btn-loading')
                  }
                >
                  Sign in
                </button>
                <button type="submit" hidden></button>
              </div>
            </div>
          </form>
        </div>
      </div>
      <style jsx>{``}</style>
    </div>
  );
};

export default Login;
