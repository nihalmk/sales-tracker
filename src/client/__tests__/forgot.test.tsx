import 'jsdom-global/register';
import React from 'react';
import { mount, shallow } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { Forgot } from './../pages/forgot';

jest.mock('@apollo/react-hooks', () => ({
  __esModule: true,
  useMutation: (): any => [
    (obj: any) => {
      const email = obj.variables.email;
      if (email === 'nosuchuser@test.com') {
        return {
          data: { sendResetPasswordEmail: false },
        };
      }
      if (email === 'john.doe@john.com') {
        return {
          data: { sendResetPasswordEmail: true },
        };
      }
    },
    {},
  ],
}));

jest.mock('next/router', () => ({
  __esModule: true,
  useRouter: () => ({
    query: {
      from: 'reset',
    },
  }),
}));

describe('Forgot password page', () => {
  let wrapper: any, email: any, sendButton: any;
  beforeEach(() => {
    wrapper = mount(<Forgot />);
    email = wrapper.find('#email').first().props();
    sendButton = wrapper.find('button').first();
  });

  it('renders as expected', async () => {
    const snapshot = shallow(<Forgot />);
    expect(snapshot).toMatchSnapshot();
  });

  it('shows an error message in case of invalid email address', async () => {
    act(() => {
      email.onChange({
        target: {
          value: 'invalid@invalid',
        },
      });
    });
    wrapper.update();
    await act(async () => {
      sendButton.simulate('click');
    });
    wrapper.update();
    const notification = wrapper.find('#notification');
    expect(notification.text()).toEqual('Invalid email address');
    expect(notification.find('.alert-danger').length).toEqual(1);
  });

  it('shows an error message when no user is found', async () => {
    act(() => {
      email.onChange({
        target: {
          value: 'nosuchuser@test.com',
        },
      });
    });

    await act(async () => {
      sendButton.simulate('click');
    });
    wrapper.update();
    const notification = wrapper.find('#notification');
    expect(notification.text()).toEqual('User not found.');
    expect(notification.find('.alert-danger').length).toEqual(1);
  });

  it('shows a success message when email is sent', async () => {
    act(() => {
      email.onChange({
        target: {
          value: 'john.doe@john.com',
        },
      } as React.ChangeEvent<HTMLInputElement>);
    });

    await act(async () => {
      sendButton.simulate('click');
    });
    wrapper.update();
    const notification = wrapper.find('#notification');
    expect(notification.text()).toEqual(
      'An email with instructions to reset password has been sent.',
    );
    expect(notification.find('.alert-success').length).toEqual(1);
  });

  it('shows the error message for expired token, when redirected from reset password page', async () => {
    wrapper = await mount(<Forgot />);
    const notification = wrapper.find('#notification');
    expect(notification.text()).toEqual(
      'Your password reset link is either invalid or expired. Create a new link below.',
    );
    expect(notification.find('.alert-danger').length).toEqual(1);
  });
});
