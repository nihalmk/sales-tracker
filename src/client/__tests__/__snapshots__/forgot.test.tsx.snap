// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forgot password page renders as expected 1`] = `
<div
  className="col col-login mx-auto"
>
  <div
    className="text-center mb-6"
  >
    <Logo
      setColor={true}
    />
  </div>
  <div
    className="card"
  >
    <div
      className="card-body"
    >
      <div
        className="card-title text-center"
      >
        Forgot Password
      </div>
    </div>
    <div
      className="card-body p-6"
    >
      <p>
        Enter your email address to receive an email with password reset instructions.
      </p>
      <div
        className="form-group"
      >
        <label
          className="form-label"
        >
          Email
        </label>
        <input
          className="form-control"
          id="email"
          name="email"
          onChange={[Function]}
          placeholder="Enter email"
          type="email"
        />
      </div>
      <div
        className="form-footer"
      >
        <button
          className="btn btn-primary btn-block false"
          onClick={[Function]}
          type="button"
        >
          Send reset email
        </button>
      </div>
    </div>
  </div>
</div>
`;
